llvm-external-projects/iree-dialects/docs/Dialect/IREEPyDM/LowerToIREE.md

```diff
@@ -0,0 +1,102 @@
+# Lowering IREEPyDM to IREE
+
+At the top-level, a program expressed in IREEPyDM ("PyDM") can be lowered to
+IREE via the overall conversion pass `-convert-iree-pydm-to-iree`; however,
+additional steps are needed to link a runtime library, etc.
+
+This document provides application notes for how the conversion is
+implemented.
+
+## Type Conversion
+
+Generally, types in PyDM map 1:1 with IREE types in a way which facilitates
+a direct conversion.
+
+## `object` and `object<type>` records
+
+Objects map to IREE variant lists (`!iree.list<?>`) of N elements, where the
+elements are laid out as:
+
+* [0]: Type Code - i32 representing the type code of the object.
+* [1] (optional): Data corresponding to the type.
+* [2] (optional): i64 TypeID - if the `id()` function has been evaluated for
+  this object, then this element will contain the unique id.
+
+### Type Codes
+
+The special type code `0` means "not assigned", which is distinct from `None`.
+Other built-in types correspond to the `BuiltinTypeCode` enum in the dialect.
+
+The following specifies the contents of the `data` field (1) of the object
+record, per type code:
+
+* `None`: No contents
+* `Tuple`: An `!iree.list` of objects
+* `List`: An `!iree.list` of objects
+* `Str`: TODO: needs to be a union of i8/i16/i32 codepoint array
+* `Bytes`: TODO: should be a memory buffer of some kind
```

Comment:
`!util.byte_buffer` exists but needs to be reworked - the `!vm.buffer` is what you want in the end and just needs to be plumbed up (same for the actual storage of Str).